File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/others/basedcyclic/SequentialCyclic.java
package cujae.inf.ic.om.heuristic.assignment.others.basedcyclic;

import java.util.ArrayList;

import cujae.inf.ic.om.heuristic.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;

import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

/*Clase que modela como asignar elementos en forma secuencial por depósitos*/
public class SequentialCyclic extends ByNotUrgency {

	public SequentialCyclic() {
		super();
	}
	
	@Override
	public Solution toClustering() {
		Solution solution = new Solution();		
		
		ArrayList<Cluster> listClusters = initializeClusters();
		ArrayList<Customer> listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		ArrayList<Integer> listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());
		NumericMatrix costMatrix = new NumericMatrix(Problem.getProblem().getCostMatrix());
		
		int totalItems = listCustomersToAssign.size();
		int totalClusters = listIDDepots.size();
		
		RowCol rcBestElement = null;
		boolean isFull = false;
		int i = 0;
		int countTry = 0;
		
		int posElementMatrix = -1;
		double capacityDepot = 0.0;
		
		int idCustomer = -1;
		double requestCustomer = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
	
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()))
		{
			posElementMatrix = totalItems + i;
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(listIDDepots.get(0)));	
			posCluster = 0;
				
			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();
			
				while(!isFull) 
				{
					rcBestElement = costMatrix.indexLowerValue(posElementMatrix, 0, posElementMatrix, (totalItems - 1));
				
					idCustomer = Problem.getProblem().getCustomers().get(rcBestElement.getCol()).getIDCustomer();
					requestCustomer = Problem.getProblem().getCustomers().get(rcBestElement.getCol()).getRequestCustomer();
					
					if(capacityDepot >= (requestCluster + requestCustomer))
					{	
						requestCluster += requestCustomer;
						
						listClusters.get(posCluster).setRequestCluster(requestCluster);
						listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

						listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));
						
						costMatrix.fillValue(rcBestElement.getRow(), 0, rcBestElement.getRow(), (totalItems - 1), Double.POSITIVE_INFINITY);
						costMatrix.fillValue(0, rcBestElement.getCol(), (totalItems + totalClusters - 1), rcBestElement.getCol(), Double.POSITIVE_INFINITY);

						posElementMatrix = Problem.getProblem().getPosElement(listClusters.get(posCluster).getItemsOfCluster().get(listClusters.get(posCluster).getItemsOfCluster().size() - 1));
						
						if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
						{
							if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
								solution.getClusters().add(listClusters.remove(posCluster));
							else
								listClusters.remove(posCluster);
							
							isFull = true;
							listIDDepots.remove(0);
						}
					} 
					else 
					{				
						costMatrix.setItem(rcBestElement.getRow(), rcBestElement.getCol(), Double.POSITIVE_INFINITY);
						countTry++;
						
						if((countTry >= listCustomersToAssign.size()))
						{
							if(!listClusters.get(posCluster).getItemsOfCluster().isEmpty())
								solution.getClusters().add(listClusters.get(posCluster));

							listClusters.remove(posCluster);
													
							listIDDepots.remove(0);
							isFull = true;
						}
					}
				}
				
				isFull = false;
				countTry = 0;	
				i++;
			}		
		}

		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		return solution;
	}		
}

File: BHAVRP_Python/cujae/inf/citi/om/heuristic/assignment/others/basedcyclic/sequential_cyclic.py
import numpy as np
from typing import List
from ...classical.by_not_urgency import ByNotUrgency
from .....problem.input.problem import Problem
from .....problem.input.customer import Customer
from .....problem.output.solution.solution import Solution
from .....problem.output.solution.cluster import Cluster

"""
Clase que modela como asignar elementos en forma secuencial por depósitos.
"""
class SequentialCyclic(ByNotUrgency):
    
    def __init__(self):
        super().__init__()
        
    def to_clustering(self) -> Solution:
        solution = Solution()
        
        # Inicialización de las listas.
        list_clusters: List[Cluster] = self.initialize_clusters()
        list_customers_to_assign: List[Customer] = list(Problem.get_problem().get_customers().copy())
        list_id_depots: List[int] = list(Problem.get_problem().get_list_id_depots())
        
        # Obtener la matriz de costos.
        cost_matrix = np.array(Problem.get_problem().get_cost_matrix())
        
        total_items = len(list_customers_to_assign)
        
        # Inicializar variables auxiliares.
        pos_element_matrix = -1
        is_full = False
        i = 0
        count_try = 0
        
        capacity_depot = 0.0
        request_customer = 0.0
        id_customer = -1
        pos_cluster = -1
        request_cluster = 0.0
        
        while list_customers_to_assign and list_clusters:
            # Escoger un depósito.
            pos_element_matrix = total_items + i
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(
                Problem.get_problem().get_depot_by_id_depot(list_id_depots[0])
            )
            pos_cluster = 0
            
            if pos_cluster != -1:
                request_cluster = list_clusters[pos_cluster].get_request_cluster()
                
                while not is_full:
                    # Encontrar el cliente con el menor costo.
                    col_index = np.argmin(cost_matrix[pos_element_matrix, :total_items])
                    
                    # Obtener la información del cliente.
                    id_customer = Problem.get_problem().get_customers()[col_index].get_id_customer()
                    request_customer = Problem.get_problem().get_customers()[col_index].get_request_customer()
                    
                    # Verificar si cabe el cliente en el depósito.
                    if capacity_depot >= (request_cluster + request_customer):
                        request_cluster += request_customer
                        
                        list_clusters[pos_cluster].set_request_cluster(request_cluster)
                        list_clusters[pos_cluster].get_items_of_cluster().append(id_customer)
                        
                        # Eliminar cliente de la lista.
                        list_customers_to_assign = [
                            customer for customer in list_customers_to_assign
                            if customer.get_id_customer() != id_customer
                        ]
                        
                        # Actualizar la matriz de costos.
                        cost_matrix[pos_element_matrix, col_index] = np.inf
                        cost_matrix[:, col_index] = np.inf
                        
                        # Verificar si el depósito está lleno.
                        if self.is_full_depot(list_customers_to_assign, request_cluster, capacity_depot):
                            if list_clusters[pos_cluster].get_items_of_cluster():
                                solution.clusters.append(list_clusters.pop(pos_cluster))
                            else:
                                list_clusters.pop(pos_cluster)
                            
                            is_full = True
                            list_id_depots.pop(0)
                    else:
                        cost_matrix[pos_element_matrix, col_index] = np.inf
                        count_try += 1
                        
                        if count_try >= len(list_customers_to_assign):
                            if list_clusters[pos_cluster].get_items_of_cluster():
                                solution.clusters.append(list_clusters[pos_cluster])
                            list_clusters.pop(pos_cluster)
                            list_id_depots.pop(0)
                            is_full = True

                # Reiniciar variables.
                is_full = False
                count_try = 0
                i += 1
                
        # Asignar clientes no asignados.
        if list_customers_to_assign:
            for customer in list_customers_to_assign:
                solution.unassigned_items.append(customer.get_id_customer())

        # Asignar clusters no vacíos.
        if list_clusters:
            for cluster in list_clusters:
                if cluster.get_items_of_cluster():
                    solution.clusters.append(cluster)
        
        return solution